gnmidiff/json.go

```diff
@@ -0,0 +1,134 @@
+// Copyright 2023 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gnmidiff
+
+import (
+	"encoding/json"
+	"fmt"
+	"sort"
+
+	"github.com/openconfig/ygot/ygot"
+)
+
+// flattenOCJSON outputs all leaf path-value pairs in the root per RFC7951.
+//
+// It assumes the input JSON is with respect to OpenConfig-compliant YANG.
+// See the following for checking compliance:
+// * https://github.com/openconfig/oc-pyang
+// * https://github.com/openconfig/public/blob/master/doc/openconfig_style_guide.md
+//
+// Output path format is per gNMI's path conventions.
+//
+// e.g.
+// {
+//   "openconfig-network-instance:config": {
+//     "description": "VRF RED",
+//     "enabled": true,
+//     "enabled-address-families": [
+//       "openconfig-types:IPV4",
+//       "openconfig-types:IPV6"
+//     ],
+//     "name": "RED",
+//     "type": "openconfig-network-instance-types:L3VRF"
+//     },
+//   "openconfig-network-instance:name": "RED"
+// }
+//
+// returns
+// {
+//   "openconfig-network-instance:config/description": "VRF RED",
+//   "openconfig-network-instance:config/enabled": true,
+//   "openconfig-network-instance:config/enabled-address-families": ["openconfig-types:IPV4", "openconfig-types:IPV6"],
+//   "openconfig-network-instance:config/name": "RED",
+//   "openconfig-network-instance:config/type": "openconfig-network-instance-types:L3VRF",
+//   "openconfig-network-instance:name": "RED",
+// }
+func flattenOCJSON(json7951 []byte) (map[string]interface{}, error) {
+	// TODO: Add option to remove the namespace on paths of returned updates.
+	var root interface{}
+	if err := json.Unmarshal(json7951, &root); err != nil {
+		return nil, fmt.Errorf("gnmidiff: %v", err)
+	}
+	leaves := map[string]interface{}{}
+	if err := flattenOCJSONAux(root, "", leaves); err != nil {
+		return nil, err
+	}
+	return leaves, nil
+}
+
+func flattenOCJSONAux(root interface{}, path string, leaves map[string]interface{}) error {
+	switch v := root.(type) {
+	case bool, float64, string:
+		leaves[path] = root
+	case []interface{}:
+		if len(v) == 0 {
```

Comment:
why isn't this just a recursive call back to itself?

Reply:
flattenOCJSONaux(root interface[], path, leaves, alreadyList)
if alreadylist return...
then handle all teh same cases?

Reply:
When `len(v) == 0` or when `v[0]` is a scalar, then this is actually a leaf-list, so it is a terminal case that needs to be handled on its own.
Otherwise, it is a list, and the recursive case needs to first handle the list key extraction before calling `flattenOCJSONAux` again.